harness/determined/_trial_controller.py

```diff
@@ -37,15 +37,16 @@ class TrialController(metaclass=abc.ABCMeta):
     def __init__(
         self,
         context: Any,
-        env: det.EnvContext,
+        env: Optional[det.EnvContext] = None,
         workloads: Optional[workload.Stream] = None,
+        det_profiler: Optional[det.profiler.ProfilerAgent] = None,
     ) -> None:
         self.context = context
         self.env = env
         # The only time that workloads should be non-None here is unit tests or test mode.
         self.workloads = workloads
 
-        self.prof = profiler.ProfilerAgent.from_env(
+        self.prof = det_profiler or env and profiler.ProfilerAgent.from_env(
```

Comment:
This is exactly the sort of complication that comes from having a bad inheritance pattern.  Inheritance should be helpful, not harmful.  This is not helping.  And it's internal-only so we can fix it however we choose.

Reply:
agreed. _trial_controller is unchanged now.